### PR TITLE
JAMES-3492 ElasticSearch 7 mailbox tests adaptation

### PR DESCRIPTION
**Pull request by LanKhuat**

### CODE REVIEW

File: mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java [hunk 1]
```diff
@@ -80,13 +91,11 @@ void tearDown() throws IOException {
         client.close();
     }
 
-    @Disabled("JAMES-3492")
 	@Override
-    protected void await() {
-        elasticSearch.awaitForElasticSearch();
+    protected void await(long numberOfResults) {
```
Comment (chibenwa) [inline mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java:95]: I propose to name this "await message count"

File: mailbox/elasticsearch-v7/src/main/java/org/apache/james/mailbox/elasticsearch/v7/search/ElasticSearchSearcher.java [hunk 2]
```diff
@@ -80,6 +81,7 @@ public ElasticSearchSearcher(ReactorElasticSearchClient client, QueryConverter q
     public Flux<MessageSearchIndex.SearchResult> search(Collection<MailboxId> mailboxIds, SearchQuery query,
                                                         Optional<Integer> limit) {
         SearchRequest searchRequest = prepareSearch(mailboxIds, query, limit);
+        System.out.println(searchRequest.toString());
```
Comment (chibenwa) [inline mailbox/elasticsearch-v7/src/main/java/org/apache/james/mailbox/elasticsearch/v7/search/ElasticSearchSearcher.java:84]: oups

File: mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java [hunk 3]
```diff
@@ -222,14 +226,13 @@ void reasonableLongTermShouldNotBeIgnored() throws Exception {
                 .setBody(reasonableLongTerm, StandardCharsets.UTF_8)),
             session).getId();
 
-        elasticSearch.awaitForElasticSearch();
+        awaitForElasticSearch(QueryBuilders.matchAllQuery(), 14);
```
Comment (chibenwa) [inline mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java:229]: Why not use awaitMessageCount ?

File: mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java [hunk 4]
```diff
@@ -286,21 +287,14 @@ void addressMatchesShouldBeExact() throws Exception {
             .setSubject("test")
             .setBody("testmail", StandardCharsets.UTF_8);
 
-        ComposedMessageId messageId1 = messageManager.appendMessage(
```
Comment (chibenwa) [inline mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java:289]: What happened to this message?

**Comment by chibenwa** [comment]
```
15:38:08,717 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (default) on project apache-james-backends-es-v7: You have 1 Checkstyle violation. -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (default) on project apache-james-backends-es-v7: You have 1 Checkstyle violation.
```

File: mailbox/elasticsearch-v7/src/main/java/org/apache/james/mailbox/elasticsearch/v7/search/ElasticSearchSearcher.java [hunk 5]
```diff
@@ -25,6 +25,7 @@
 import java.util.Optional;
 
 import org.apache.james.backends.es.v7.AliasName;
+import org.apache.james.backends.es.v7.NodeMappingFactory;
```
Comment (Arsnael) [inline mailbox/elasticsearch-v7/src/main/java/org/apache/james/mailbox/elasticsearch/v7/search/ElasticSearchSearcher.java:28]: what is this import for?

File: backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/NodeMappingFactory.java [hunk 6]
```diff
@@ -54,15 +51,18 @@
...
         return client;
     }
 
+    //TODO: since ElasticSearch version 7 this check does not work anymore. Find another way to check whether mapping exists or not
     // ElasticSearch 6.3.2 does not support field master_timeout that is set up by 6.4.3 REST client when relying on getMapping
     @SuppressWarnings("deprecation")
-    public static boolean mappingAlreadyExist(ReactorElasticSearchClient client, IndexName indexName) throws IOException {
+    /*public static boolean mappingAlreadyExist(ReactorElasticSearchClient client, IndexName indexName) throws IOException {
         try {
             client.getLowLevelClient().performRequest(new Request("GET", "/" + indexName.getValue() + "/_mapping/"));
```
Comment (Arsnael) [inline backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/NodeMappingFactory.java:67]: https://www.elastic.co/guide/en/elasticsearch/reference/7.10/indices-get-mapping.html#get-mapping-api-query-params

maybe trying with putting `allow_no_indices` param to false?

On an other note as well, before the client for 6.3 didn't have a method to get the mapping of an indice, which is why we used the low level API. The 7.10 has a method it seems though: https://www.elastic.co/guide/en/elasticsearch/client/java-rest/7.10/java-rest-high-get-mappings.html

You might want to check it out too :)

File: mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java (hunk 4 above)
Comment (LanKhuat) [reply mailbox/elasticsearch-v7/src/test/java/org/apache/james/mailbox/elasticsearch/v7/ElasticSearchIntegrationTest.java]: deleted by mistake, sorry

**Review by Arsnael — APPROVED** [review]

File: backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/NodeMappingFactory.java [hunk 7]
```diff
@@ -60,18 +59,18 @@ public static ReactorElasticSearchClient applyMapping(ReactorElasticSearchClient
         return client;
     }
 
-    // ElasticSearch 6.3.2 does not support field master_timeout that is set up by 6.4.3 REST client when relying on getMapping
-    @SuppressWarnings("deprecation")
     public static boolean mappingAlreadyExist(ReactorElasticSearchClient client, IndexName indexName) throws IOException {
```
Comment (chibenwa) [inline backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/NodeMappingFactory.java:62]: Where is this method used? Do we need it in production code?

File: backends-common/elasticsearch-v7/src/test/java/org/apache/james/backends/es/v7/ElasticSearchIndexerTest.java [hunk 8]
```diff
@@ -69,7 +70,7 @@ void setup() {
         new IndexCreationFactory(ElasticSearchConfiguration.DEFAULT_CONFIGURATION)
             .useIndex(INDEX_NAME)
             .addAlias(ALIAS_NAME)
-            .createIndexAndAliases(client);
+            .createIndexAndAliases(client, Optional.empty());
```
Comment (chibenwa) [inline backends-common/elasticsearch-v7/src/test/java/org/apache/james/backends/es/v7/ElasticSearchIndexerTest.java:73]: What does `Optional.empty()` means?

File: backends-common/elasticsearch-v7/src/test/java/org/apache/james/backends/es/v7/IndexCreationFactoryTest.java [hunk 9]
```diff
@@ -42,7 +43,7 @@ void setUp() {
         new IndexCreationFactory(ElasticSearchConfiguration.DEFAULT_CONFIGURATION)
             .useIndex(INDEX_NAME)
             .addAlias(ALIAS_NAME)
-            .createIndexAndAliases(client);
+            .createIndexAndAliases(client, Optional.empty());
```
Comment (chibenwa) [inline backends-common/elasticsearch-v7/src/test/java/org/apache/james/backends/es/v7/IndexCreationFactoryTest.java:46]: Idem

File: backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/NodeMappingFactory.java (hunk 7 above)
Comment (LanKhuat) [reply backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/NodeMappingFactory.java:62]: i don't think so. Only used in tests now. Should we remove this class along with all its tests?

Comment (chibenwa) [reply backends-common/elasticsearch-v7/src/main/java/org/apache/james/backends/es/v7/NodeMappingFactory.java:62]: IMO yes

File: backends-common/elasticsearch-v7/src/test/java/org/apache/james/backends/es/v7/NodeMappingFactoryTest.java [hunk 10]
```diff
@@ -46,7 +46,7 @@ void setUp() throws Exception {
         new IndexCreationFactory(ElasticSearchConfiguration.DEFAULT_CONFIGURATION)
             .useIndex(INDEX_NAME)
             .addAlias(ALIAS_NAME)
-            .createIndexAndAliases(client, Optional.of(getMappingsSources()));
+            .createIndexAndAliasesWithMapping(client, Optional.of(getMappingsSources()));
```
Comment (chibenwa) [inline backends-common/elasticsearch-v7/src/test/java/org/apache/james/backends/es/v7/NodeMappingFactoryTest.java]: Then this argument should not be an option, don't you think?

**Review by chibenwa — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**